src/policy/policy.cpp

```diff
@@ -231,6 +231,25 @@ bool IsWitnessStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
                     return false;
             }
         }
+
+        // Check P2TR standard limits
+        if (witnessversion == 1 && witnessprogram.size() == WITNESS_V1_TAPROOT_SIZE && !prevScript.IsPayToScriptHash()) {
+            // Taproot spend
+            const auto& stack = tx.vin[i].scriptWitness.stack;
+            size_t stack_size = stack.size();
+            if (stack_size >= 2 && !stack[stack_size - 1].empty() && stack[stack_size - 1][0] == ANNEX_TAG) {
+                stack_size--; // Ignore annex
+            }
+            if (stack_size >= 2) {
+                // Script path spend
+                if ((stack[stack_size - 1][0] & 0xfe) == 0xc0) {
```

Comment:
why nested conditional here, but not nested above (line 240)?  both branches have "stack_size >= 2" as well. without understanding what any of this means, would it be clearer to write like this?

```
// precondition for taproot (?)
if (stack_size < 2) { return }

if (!stack[stack_size - 1].empty() && stack[stack_size - 1][0] == ANNEX_TAG) {
// whatever this is

} else if ((stack[stack_size - 1][0] & 0xfe) == 0xc0)) {
// script path spend

}
```

Reply:
@skwp I'll add some comments to clarify, but what you're suggesting would be incorrect.

Reply:
Thanks @sipa. I'm not trying to nitpick, rather looking at ways to lower the barrier for reviewers and thus help get more eyes on things. I find that complex conditionals with lots of low level details can make bugs hard to spot and make it daunting for people new to the source to contribute. There might be good reason for the style employed throughout the code and we probably can't change everything overnight, but even things like magic numbers (0xfe?? 0xc0??) everywhere makes it really hard to reason about, imho. Seems like we can make the source easier to read as we go with named conditionals, constants, etc.

Reply:
In fact I see that these magic numbers are already given names elsewhere. Can we use them here? This is way more readable and expresses intent

```
(stack[stack_size - 1][0] & TAPROOT_LEAF_MASK) == TAPROOT_LEAF_TAPSCRIPT
```

Reply:
Changed it to use the constants, that's obviously the better way.

The reason why the annex does not use a nested conditional is because it's a single `if` statement that tests the presence of the annex (as in: both conditions are needed to have an annex). The tapscript conditional is split up into the logical "this is a script path spending" condition and the "this is a leaf version 0xc0 script".